src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -5,73 +5,223 @@
 using Microsoft.AspNetCore.Razor.LanguageServer.Semantic.Services;
+using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using OmniSharp.Extensions.LanguageServer.Protocol.Models;
 using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
 {
     internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoService
     {
-        // This cache is not created for performance, but rather to restrict memory growth.
+        // This cache is created to restrict memory growth.
         // We need to keep track of the last couple of requests for use in previousResultId,
         // but if we let the grow unbounded it could quickly allocate a lot of memory.
         // Solution: an in-memory cache
-        private static readonly MemoryCache<string, IReadOnlyList<int>> _semanticTokensCache =
-            new MemoryCache<string, IReadOnlyList<int>>();
+        private readonly MemoryCache<string, (VersionStamp, IReadOnlyList<int>)> _semanticTokensCache =
+            new MemoryCache<string, (VersionStamp, IReadOnlyList<int>)>();
 
-        public override SemanticTokens GetSemanticTokens(RazorCodeDocument codeDocument)
+        private readonly IClientLanguageServer _languageServer;
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public DefaultRazorSemanticTokensInfoService(
+            IClientLanguageServer languageServer,
+            RazorDocumentMappingService documentMappingService)
+        {
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        public override Task<SemanticTokens> GetSemanticTokensAsync(DocumentSnapshot documentSnapshot, TextDocumentIdentifier textDocumentIdentifier, CancellationToken cancellationToken)
         {
-            return GetSemanticTokens(codeDocument, range: null);
+            return GetSemanticTokensAsync(documentSnapshot, textDocumentIdentifier, range: null, cancellationToken);
         }
 
-        public override SemanticTokens GetSemanticTokens(RazorCodeDocument codeDocument, Range range)
+        public override async Task<SemanticTokens> GetSemanticTokensAsync(
+            DocumentSnapshot documentSnapshot,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range range,
+            CancellationToken cancellationToken)
         {
+            var codeDocument = await GetRazorCodeDocument(documentSnapshot);
             if (codeDocument is null)
             {
                 throw new ArgumentNullException(nameof(codeDocument));
             }
+            var documentVersion = await documentSnapshot.GetTextVersionAsync();
 
```

Comment:
should add a cancellationToken.ThrowIfCanclelationREquested() here since it's inbetween awaits. Ditto elsewhere in your changes